mino/minows/rpc.go

```diff
@@ -0,0 +1,422 @@
+			return xerrors.Errorf("could not serialize message: %v", err)
+		}
+		payload = bytes
+	}
+
+	err := enc.Encode(&Packet{Source: from, Payload: payload})
+	if errors.Is(err, network.ErrReset) {
+		return err
+	}
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	return nil
+}
+
+func (r rpc) receive(dec *gob.Decoder) (ma.Multiaddr, serde.Message, error) {
+	var packet Packet
+	err := dec.Decode(&packet)
+	if errors.Is(err, network.ErrReset) {
+		return nil, nil, err
+	}
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not decode packet: %v", err)
+	}
+
+	from, err := ma.NewMultiaddrBytes(packet.Source)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not unmarshal address: %v",
+			packet.Source)
+	}
+
+	if packet.Payload == nil {
+		return from, nil, nil
+	}
+	msg, err := r.factory.Deserialize(r.context, packet.Payload)
+	if err != nil {
+		return from, nil, xerrors.Errorf(
+			"could not deserialize message: %v",
+			err)
+	}
+	return from, msg, nil
+}
+
+func (r rpc) createOrchestrator(ctx context.Context,
+	initiator host.Host, streams []network.Stream) (*orchestrator, error) {
+	participant := r.mino.GetAddress().(address).location
+	myAddr, err := newOrchestratorAddr(participant, initiator.ID())
+	if err != nil {
+		return nil, xerrors.Errorf("could not create address: %v", err)
+	}
+
+	encoders := make(map[peer.ID]*gob.Encoder)
+	for _, stream := range streams {
+		encoders[stream.Conn().RemotePeer()] = gob.NewEncoder(stream)
+	}
+
+	o := &orchestrator{
+		logger: r.logger.With().Stringer("mino", myAddr).
+			Stringer("orchestrator", xid.New()).Logger(),
+		myAddr: myAddr,
+		rpc:    r,
+		outs:   encoders,
+		in:     make(chan Packet, MaxUnreadAllowed),
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(streams))
+	for _, stream := range streams {
+		go func(stream network.Stream) {
+			defer wg.Done()
+			decoder := gob.NewDecoder(stream)
+			for {
+				packet, err := o.listen(decoder)
+				if err != nil {
+					if strings.Contains(err.Error(), network.ErrReset.Error()) {
+						return
+					}
+					r.logger.Error().Err(err).Msg("message dropped")
+					continue
+				}
+				select {
+				case <-ctx.Done():
+					return
+				case o.in <- packet:
+				}
+			}
+		}(stream)
+	}
+
+	go func() {
+		wg.Wait()
+		close(o.in)
+	}()
+
+	return o, nil
+}
+
+func (r rpc) createParticipant(stream network.Stream) participant {
+	encoder := gob.NewEncoder(stream)
+	decoder := gob.NewDecoder(stream)
+
+	p := participant{
+		logger: r.logger.With().Stringer("participant", xid.New()).Logger(),
+		myAddr: r.mino.myAddr,
+		rpc:    r,
+		out:    encoder,
+		in:     make(chan Packet),
+	}
+
+	done := make(chan any)
+	go func() {
+		hasReset := func() bool {
+			for _, s := range stream.Conn().GetStreams() {
+				if s.ID() == stream.ID() {
+					return false
+				}
+			}
+			return true
+		}
+
+		for !hasReset() {
+			time.Sleep(2 * time.Second)
+		}
+		close(done)
+	}()
+
+	go func() {
+		for {
```

Comment:
this go routine is virtually a copy-paste from the orchestrator, reinforcing the idea it should be its own separate function.